File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] + [amount + 1] * (amount+1)
        
        for i in range(1, amount + 1):
            for coin in coins:
                if i-coin>=0:
                    dp[i] = min(dp[i], dp[i - coin] + 1)

        return -1 if dp[amount] == amount + 1 else dp[amount]

'''
    [1,2,5], amount = 11
    1-10, 1-9

    select 1 - DFS
'''

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, -1);
        dp[0] = 0;

        for(int i=1; i<=amount; i++) {
            for(int c : coins) {
                if(i-c>=0 && dp[i-c]!= -1){
                    dp[i] = dp[i]>0? min(dp[i], dp[i-c]+1) : dp[i-c]+1;
                }
            }
        }
        return dp[amount];
    }
};

/*
    coins = [1,2,5], amount = 11

    dp 0   1   2   3
       0  -1. -1. -1
                        7
                6       5         2
             5 4 1.  4  3 0.   1. 0 -3
        1 3 0 
            1. 2  5. 1.  2. 5  1  2  5


*/